packages/layout/src/FlexGrid.ts

```diff
@@ -0,0 +1,139 @@
+import { HTMLWidget } from "@hpcc-js/common";
+
+import "../src/FlexGrid.css";
+
+export class FlexGrid extends HTMLWidget {
+    constructor() {
+        super();
+    }
+    enter(domNode, element) {
+        super.enter(domNode, element);
+    }
+    update(domNode, element) {
+        super.update(domNode, element);
+        const context = this;
+        this.resize();
+
+        const cachedSizes = [];
+
+        this.updateFlexParent(element);
+        const listItems = element.selectAll(".FlexGrid-list-item").data(this.widgets(), w => w.id());
+        listItems.enter()
+            .append("div")
+            .classed("FlexGrid-list-item", true)
+            .style("height", this.itemMinHeight() + "px")
+            .style("width", this.itemMinWidth() + "px")
+            .each(function(w, i) {
+                w.target(this);
+            })
+            .merge(listItems)
+            .style("height", null)
+            .style("width", null)
+            .style("min-height", this.itemMinHeight() + "px")
+            .style("min-width", this.itemMinWidth() + "px")
+            .style("flex-basis", (n, i) => {
+                if (this.widgetsFlexBasis()[i]) {
+                    return this.widgetsFlexBasis()[i];
+                }
+                return this.flexBasis();
+            })
+            .style("flex-grow", (n, i) => {
+                if (this.widgetsFlexGrow()[i]) {
+                    return this.widgetsFlexGrow()[i];
+                }
+                return this.flexGrow();
+            })
+            .style("border-width", this.borderWidth() + "px")
+            .style("border-color", this.borderColor())
+            .each(function() {
+                this.firstChild.style.display = "none";
+            })
+            .each(function() {
+                const rect = this.getBoundingClientRect();
+                cachedSizes.push([
+                    rect.width,
+                    rect.height
+                ]);
+            })
+            .each(function(w, i) {
+                this.firstChild.style.display = "block";
+                w.resize({
+                    width: cachedSizes[i][0] - (2 * context.borderWidth()),
+                    height: cachedSizes[i][1] - (2 * context.borderWidth())
+                });
+            })
```

Comment:
I am not sure what the 3 `each` are trying to do?

Reply:
1. The first `each` sets `display: none`. This prevents flex child elements from being affected by the dimensions of the inserted widgets.
2. The second `each` caches the dimensions of the all the flex child elements. Inserting the widgets at this point would cause all the other flex child element dimensions to change after each insertion.
3. The third `each` applies the cached dimensions.